webdriver/support/fixtures.py

```diff
@@ -83,54 +88,88 @@ def create_window():
-        request.addfinalizer(destroy)
-        request.addfinalizer(lambda: _switch_to_top_level_browsing_context(session))
-        request.addfinalizer(lambda: _restore_windows(session))
-        request.addfinalizer(lambda: _dismiss_user_prompts(session))
-        request.addfinalizer(lambda: _ensure_valid_window(session))
-
-        return session
+
+def configuration():
+    host = os.environ.get("WD_HOST", default_host)
+    port = int(os.environ.get("WD_PORT", default_port))
+    capabilities = json.loads(os.environ.get("WD_CAPABILITIES", "{}"))
+
+    return {
+        "host": host,
+        "port": port,
+        "capabilities": capabilities
+    }
+
+
+_current_session = None
+
+
+def session(configuration, request):
+    """Create and start a session for a test that does not itself test session creation.
+
+    By default the session will stay open after each test, but we always try to start a
+    new one and assume that if that fails there is already a valid session. This makes it
+    possible to recover from some errors that might leave the session in a bad state, but
+    does not demand that we start a new session per test."""
+    global _current_session
+    if _current_session is None:
+        _current_session = webdriver.Session(configuration["host"],
+                                             configuration["port"],
+                                             capabilities=configuration["capabilities"])
+    try:
+        _current_session.start()
+    except webdriver.errors.SessionNotCreatedException:
+        if not _current_session.session_id:
+            raise
+
+    # finalisers are popped off a stack,
+    # making their ordering reverse
+    request.addfinalizer(lambda: _switch_to_top_level_browsing_context(_current_session))
+    request.addfinalizer(lambda: _restore_windows(_current_session))
+    request.addfinalizer(lambda: _dismiss_user_prompts(_current_session))
+    request.addfinalizer(lambda: _ensure_valid_window(_current_session))
+
+    return _current_session
+
+
+def new_session(configuration, request):
+    """Return a factory function that will attempt to start a session with a given body.
+
+    This is intended for tests that are themselves testing new session creation, and the
+    session created is closed at the end of the test."""
+    def end():
+        global _current_session
+        if _current_session is not None and _current_session.session_id:
+            _current_session.end()
+            _current_session = None
+
+    def create_session(body):
+        global _current_session
+        _session = webdriver.Session(configuration["host"],
+                                     configuration["port"],
+                                     capabilities=None)
+        # TODO: merge in some capabilities from the confguration capabilities
+        # since these might be needed to start the browser
+        value = _session.send_command("POST", "session", body=body)
```

Comment:
> I think making it work well might require splitting the capabilites we store
> into "required" and "optional" or something, where "required" would probably
> be just the path to the binary and optional would be anything like log level
> that isn't really needed to start a successful session.

The `merge_dictionaries` utility function gives us this functionality because it throws when the second dictionary re-defines a key that is present in the first.

> Just returning the session object doesn't seem like a big win to me? We
> probably still need the rest of the logic here but it would end up in a
> function in the tests.

The change I have in mind is minimal and would only result in one extra line of code within the tests, `session.start()` (more on this in a moment):

```diff
 def create_session(body):
     global _current_session
     _session = webdriver.Session(configuration["host"],
                                  configuration["port"],
                                  capabilities=None)
     # TODO: merge in some capabilities from the confguration capabilities
     # since these might be needed to start the browser
-    value = _session.send_command("POST", "session", body=body)
     # Don't set the global session until we are sure this succeeded
     _current_session = _session
-    _session.session_id = value["sessionId"]
 
-    return value, _current_session
+    return _current_session
```

> In any case the tests in #5187 depend on this API and I don't think they're
> too confusing.

The response body is really only useful for those tests. The majority of tests that rely on this fixture will have to shunt that value.

Of course, if we *don't* start the session in the fixture, those tests will have to do that themselves (as mentioned above, via `session.start()`. This is a different (and longer) kind of repetition. But it's also in keeping with the approach we agreed on a while back: tests ought to use high-level, exception-raising methods for details that they are not directly validating. For functionality under test, though, tests really ought to include the request-response handling. This would make [your new tests](https://github.com/w3c/web-platform-tests/pull/5187) a little more coherent to the average contributor: instead of asserting that "calling this Pytest fixture function raises a wdclient exception", the tests would say "sending a New Session command with wdclient produces an HTTP response with such-and-such status code and content."